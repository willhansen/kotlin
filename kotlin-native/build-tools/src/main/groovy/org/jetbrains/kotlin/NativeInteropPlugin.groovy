/*
 * Copyright 2010-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin

import org.gradle.api.Named
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.artifacts.Configuration
import org.gradle.api.file.FileCollection
import org.gradle.api.internal.AbstractNamedDomainObjectContainer
import org.gradle.api.internal.CollectionCallbackActionDecorator
import org.gradle.api.tasks.JavaExec
import org.gradle.api.tasks.SourceSet
import org.gradle.internal.reflect.Instantiator

class NamedNativeInteropConfig implements Named {

    private final Project project
    final String name


    private String interopStubsName
    private SourceSet interopStubs
    final JavaExec genTask

    private String flavor = "jvm"

    private String defFile

    private String pkg
    private String target

    private List<String> compilerOpts = []
    private List<String> headers;
    private String linker
    List<String> linkerOpts = []
    private FileCollection linkFiles;
    private List<String> linkTasks = []

    Configuration configuration

    void flavor(String konstue) {
        flavor = konstue
    }

    void defFile(String konstue) {
        defFile = konstue
        genTask.inputs.file(project.file(defFile))
    }

    void target(String konstue) {
        target = konstue
    }

    void pkg(String konstue) {
        pkg = konstue
    }

    void compilerOpts(List<String> konstues) {
        compilerOpts.addAll(konstues)
    }

    void compilerOpts(String... konstues) {
        compilerOpts.addAll(konstues)
    }

    void headers(FileCollection files) {
        dependsOnFiles(files)
        headers = headers + files.toSet().collect { it.absolutePath }
    }

    void headers(String... konstues) {
        headers = headers + konstues.toList()
    }

    void linker(String konstue) {
        linker = konstue
    }

    void linkerOpts(String... konstues) {
        this.linkerOpts(konstues.toList())
    }

    void linkerOpts(List<String> konstues) {
        linkerOpts.addAll(konstues)
    }

    void dependsOn(Object... deps) {
        // TODO: add all files to inputs
        genTask.dependsOn(deps)
    }

    private void dependsOnFiles(FileCollection files) {
        dependsOn(files)
        genTask.inputs.files(files)
    }

    void link(FileCollection files) {
        linkFiles = linkFiles + files
        dependsOnFiles(files)
    }

    void linkOutputs(Task task) {
        linkOutputs(task.name)
    }

    void linkOutputs(String task) {
        linkTasks += task
        dependsOn(task)

        final Project prj;
        String taskName;
        int index = task.lastIndexOf(':')
        if (index != -1) {
            prj = project.project(task.substring(0, index))
            taskName = task.substring(index + 1)
        } else {
            prj = project
            taskName = task
        }

        prj.tasks.matching { it.name == taskName }.all { // TODO: it is a hack
            this.dependsOnFiles(it.outputs.files)
        }
    }

    void includeDirs(String... konstues) {
        compilerOpts.addAll(konstues.collect {"-I$it"})
    }

    File getNativeLibsDir() {
        return new File(project.buildDir, "nativelibs/$target")
    }

    File getGeneratedSrcDir() {
        return new File(project.buildDir, "nativeInteropStubs/$name/kotlin")
    }

    File getTemporaryFilesDir() {
        return new File(project.buildDir, "interopTemp")
    }

    NamedNativeInteropConfig(Project project, String name, String target = null, String flavor = 'jvm') {
        this.name = name
        this.project = project
        this.flavor = flavor

        def platformManager = project.project(":kotlin-native").ext.platformManager
        def targetManager = platformManager.targetManager(target)
        this.target = targetManager.targetName

        this.headers = []
        this.linkFiles = project.files()

        interopStubsName = name + "InteropStubs"
        genTask = project.task("gen" + interopStubsName.capitalize(), type: JavaExec)

        this.configure()
    }

    private void configure() {
        if (project.plugins.hasPlugin("kotlin")) {
            interopStubs = project.sourceSets.create(interopStubsName)
            configuration = project.configurations.create(interopStubs.name)
            project.tasks.getByName(interopStubs.getTaskName("compile", "Kotlin")) {
                dependsOn genTask
                kotlinOptions.freeCompilerArgs = ["-Xskip-prerelease-check"]
            }

            interopStubs.kotlin.srcDirs generatedSrcDir

            project.dependencies {
                add interopStubs.getApiConfigurationName(), project(path: ':kotlin-native:Interop:Runtime')
                add interopStubs.getApiConfigurationName(), "org.jetbrains.kotlin:kotlin-stdlib:${project.bootstrapKotlinVersion}"
            }

            this.configuration.extendsFrom project.configurations[interopStubs.getRuntimeClasspathConfigurationName()]
            project.dependencies.add(this.configuration.name, interopStubs.output)
        }

        genTask.configure {
            dependsOn ":kotlin-native:dependencies:update"
            dependsOn ":kotlin-native:Interop:Indexer:nativelibs"
            dependsOn ":kotlin-native:Interop:Runtime:nativelibs"
            classpath = project.configurations.interopStubGenerator
            mainClass = "org.jetbrains.kotlin.native.interop.gen.jvm.MainKt"
            jvmArgs '-ea'

            systemProperties "java.library.path" : project.files(
                    new File(project.findProject(":kotlin-native:Interop:Indexer").buildDir, "nativelibs"),
                    new File(project.findProject(":kotlin-native:Interop:Runtime").buildDir, "nativelibs")
            ).asPath
            // Set the konan.home property because we run the cinterop tool not from a distribution jar
            // so it will not be able to determine this path by itself.
            systemProperties "konan.home": project.project(":kotlin-native").projectDir
            environment "LIBCLANG_DISABLE_CRASH_RECOVERY": "1"

            outputs.dir generatedSrcDir
            outputs.dir nativeLibsDir
            outputs.dir temporaryFilesDir

            // defer as much as possible
            doFirst {
                List<String> linkerOpts = this.linkerOpts

                linkTasks.each {
                    linkerOpts += project.tasks.getByPath(it).outputs.files.files
                }

                linkerOpts += linkFiles.files

                args '-generated', generatedSrcDir
                args '-natives', nativeLibsDir
                args '-Xtemporary-files-dir', temporaryFilesDir
                args '-flavor', this.flavor
                if (flavor == "jvm") {
                    args '-mode', 'sourcecode'
                }
                // Uncomment to debug.
                // args '-verbose', 'true'

                if (defFile != null) {
                    args '-def', project.file(defFile)
                }

                if (pkg != null) {
                    args '-pkg', pkg
                }

                if (linker != null) {
                    args '-linker', linker
                }

                if (target != null) {
                    args '-target', target
                }

                // TODO: the interop plugin should probably be reworked to execute clang from build scripts directly
                environment['PATH'] = project.files(project.extensions.platformManager.hostPlatform.clang.clangPaths).asPath +
                        File.pathSeparator + environment['PATH']

                args compilerOpts.collectMany { ['-compiler-option', it] }
                args linkerOpts.collectMany { ['-linker-option', it] }

                headers.each {
                    args '-header', it
                }

            }
        }
    }
}

class NativeInteropExtension extends AbstractNamedDomainObjectContainer<NamedNativeInteropConfig> {

    private final Project project
    private String target = null
    private String flavor = 'jvm'

    protected NativeInteropExtension(Project project) {
        super(NamedNativeInteropConfig, project.gradle.services.get(Instantiator), project.gradle.services.get(CollectionCallbackActionDecorator))
        this.project = project
    }

    @Override
    protected NamedNativeInteropConfig doCreate(String name) {
        def config = new NamedNativeInteropConfig(project, name, target, flavor)
        return config
    }

    public void target(String konstue) {
        this.target = konstue
    }

    public void flavor(String konstue) {
        this.flavor = konstue
    }
}

class NativeInteropPlugin implements Plugin<Project> {

    @Override
    void apply(Project prj) {
        prj.extensions.add("kotlinNativeInterop", new NativeInteropExtension(prj))

        prj.configurations {
            interopStubGenerator
        }

        prj.dependencies {
            interopStubGenerator project(path: ":kotlin-native:Interop:StubGenerator")
            interopStubGenerator project(path: ":kotlin-native:endorsedLibraries:kotlinx.cli", configuration: "jvmRuntimeElements")
        }
    }
}
