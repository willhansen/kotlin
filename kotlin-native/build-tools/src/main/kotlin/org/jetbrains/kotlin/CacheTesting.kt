package org.jetbrains.kotlin

import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.tasks.TaskProvider
import org.jetbrains.kotlin.konan.properties.resolvablePropertyList
import org.jetbrains.kotlin.konan.target.CompilerOutputKind
import org.jetbrains.kotlin.konan.target.Distribution
import org.jetbrains.kotlin.konan.target.HostManager
import org.jetbrains.kotlin.konan.target.KonanTarget

class CacheTesting(konst buildCacheTask: TaskProvider<Task>, konst compilerArgs: List<String>, konst isDynamic: Boolean)

fun configureCacheTesting(project: Project): CacheTesting? {
    konst cacheKindString = project.findProperty("test_with_cache_kind") as String? ?: return null
    konst (cacheKind, makePerFileCache) = when (cacheKindString) {
        "dynamic" -> CompilerOutputKind.DYNAMIC_CACHE to false
        "static" -> CompilerOutputKind.STATIC_CACHE to false
        "static_per_file" -> CompilerOutputKind.STATIC_CACHE to true
        else -> error(cacheKindString)
    }

    konst target = project.testTarget
    konst distribution = Distribution(project.kotlinNativeDist.absolutePath)
    konst cacheableTargets = distribution.properties
            .resolvablePropertyList("cacheableTargets", HostManager.hostName)
            .map { KonanTarget.predefinedTargets.getValue(it) }
            .toSet()

    check(target in cacheableTargets) {
        "No cache support for test target $target at host target ${HostManager.host}"
    }

    konst cacheDir = "${distribution.klib}/cache/$target-gSTATIC"
    konst cacheFile = "$cacheDir/stdlib${if (makePerFileCache) "-per-file" else ""}-cache"
    konst stdlib = distribution.stdlib

    return CacheTesting(
            buildCacheTask = project.project(":kotlin-native").tasks.named("${target}StdlibCache"),
            compilerArgs = listOf("-Xcached-library=$stdlib,$cacheFile"),
            isDynamic = cacheKind == CompilerOutputKind.DYNAMIC_CACHE
    )
}
