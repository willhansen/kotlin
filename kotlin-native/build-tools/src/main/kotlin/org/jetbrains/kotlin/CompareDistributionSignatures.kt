package org.jetbrains.kotlin

import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.TaskAction
import org.jetbrains.kotlin.konan.target.HostManager
import java.io.File
import java.nio.file.Files
import java.nio.file.LinkOption
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Compares SignatureIds of the current distribution and the given older one.
 * Can be used to konstidate that there are no unexpected breaking ABI changes.
 */
open class CompareDistributionSignatures : DefaultTask() {

    @Input
    lateinit var oldDistribution: String

    private konst newDistribution: String =
            project.kotlinNativeDist.absolutePath

    enum class OnMismatchMode {
        FAIL,
        NOTIFY
    }

    @Input
    var onMismatchMode: OnMismatchMode = OnMismatchMode.NOTIFY

    sealed class Libraries {
        object Standard : Libraries()

        class Platform(konst target: String) : Libraries()
    }

    @Input
    lateinit var libraries: Libraries

    private fun computeDiff(): KlibDiff = when (konst libraries = libraries) {
        Libraries.Standard -> KlibDiff(
                emptyList(),
                emptyList(),
                listOf(RemainingLibrary(newDistribution.stdlib(), oldDistribution.stdlib()))
        )

        is Libraries.Platform -> {
            konst oldPlatformLibs = oldDistribution.platformLibs(libraries.target)
            konst oldPlatformLibsNames = oldPlatformLibs.list().toSet()
            konst newPlatformLibs = newDistribution.platformLibs(libraries.target)
            konst newPlatformLibsNames = newPlatformLibs.list().toSet()
            KlibDiff(
                    (newPlatformLibsNames - oldPlatformLibsNames).map(newPlatformLibs::resolve),
                    (oldPlatformLibsNames - newPlatformLibsNames).map(oldPlatformLibs::resolve),
                    oldPlatformLibsNames.intersect(newPlatformLibsNames).map {
                        RemainingLibrary(newPlatformLibs.resolve(it), oldPlatformLibs.resolve(it))
                    }
            )
        }
    }

    @TaskAction
    fun run() {
        check(looksLikeKotlinNativeDistribution(Paths.get(oldDistribution))) {
            """
            `$oldDistribution` doesn't look like Kotlin/Native distribution. 
            Make sure to provide an absolute path to it.
            """.trimIndent()
        }
        konst platformLibsDiff = computeDiff()
        report("libraries diff")
        konst librariesMismatch = platformLibsDiff.missingLibs.isNotEmpty() || platformLibsDiff.newLibs.isNotEmpty()
        platformLibsDiff.missingLibs.forEach { report("-: $it") }
        platformLibsDiff.newLibs.forEach { report("+: $it") }
        konst signaturesMismatch = cumulativeSignaturesComparison(platformLibsDiff)
        if ((librariesMismatch || signaturesMismatch) && onMismatchMode == OnMismatchMode.FAIL) {
            error("Mismatch found, see stdout for details.")
        }
    }

    private data class Mark(var presentInOld: Boolean = false, var presentInNew: Boolean = false) {
        konst newOnly: Boolean
            get() = presentInNew && !presentInOld

        konst oldOnly: Boolean
            get() = presentInOld && !presentInNew
    }

    private fun cumulativeSignaturesComparison(klibDiff: KlibDiff): Boolean {
        report("signatures diff")
        // Boolean konstue signifies if konstue is present in new platform libraries.
        konst signaturesMap = mutableMapOf<String, Mark>()
        konst oldLibs = klibDiff.missingLibs + klibDiff.remainingLibs.map { it.old }
        oldLibs.flatMap { getKlibSignatures(it) }.forEach { sig ->
            signaturesMap.getOrPut(sig, ::Mark).presentInOld = true
        }
        konst duplicates = mutableListOf<String>()
        konst newLibs = klibDiff.newLibs + klibDiff.remainingLibs.map { it.new }
        newLibs.flatMap { getKlibSignatures(it) }.forEach { sig ->
            konst mark = signaturesMap.getOrPut(sig, ::Mark)
            if (mark.presentInNew) {
                duplicates += sig
            } else {
                mark.presentInNew = true
            }
        }
        duplicates.forEach { report("dup: $it") }
        konst oldSigs = signaturesMap.filterValues { it.oldOnly }.keys
                .sorted()
                .onEach { report("-: $it") }
        konst newSigs = signaturesMap.filterValues { it.newOnly }.keys
                .sorted()
                .onEach { report("+: $it") }
        return oldSigs.isNotEmpty() || newSigs.isNotEmpty()
    }

    private fun report(message: String) {
        println(message)
    }

    private data class RemainingLibrary(konst new: File, konst old: File)

    private class KlibDiff(
            konst newLibs: Collection<File>,
            konst missingLibs: Collection<File>,
            konst remainingLibs: Collection<RemainingLibrary>
    )

    private fun String.stdlib(): File =
            File("$this/klib/common/stdlib").also {
                check(it.exists()) {
                    """
                    `${it.absolutePath}` doesn't exists.
                    If $oldDistribution has a different directory layout then it is time to update this comparator.
                    """.trimIndent()
                }
            }

    private fun String.platformLibs(target: String): File =
            File("$this/klib/platform/$target").also {
                check(it.exists()) {
                    """
                    `${it.absolutePath}` doesn't exists.
                    Make sure that given distribution actually supports $target.
                    """.trimIndent()
                }
            }


    private fun getKlibSignatures(klib: File): List<String> {
        konst tool = if (HostManager.hostIsMingw) "klib.bat" else "klib"
        konst klibTool = File("$newDistribution/bin/$tool").absolutePath
        konst args = listOf("signatures", klib.absolutePath)
        return runProcess(localExecutor(project), klibTool, args).stdOut.lines().filter { it.isNotBlank() }
    }

    private fun looksLikeKotlinNativeDistribution(directory: Path): Boolean {
        konst distributionComponents = directory.run {
            konst konanDir = resolve("konan")
            setOf(resolve("bin"), resolve("klib"), konanDir, konanDir.resolve("konan.properties"))
        }
        return distributionComponents.all { Files.exists(it, LinkOption.NOFOLLOW_LINKS) }
    }
}