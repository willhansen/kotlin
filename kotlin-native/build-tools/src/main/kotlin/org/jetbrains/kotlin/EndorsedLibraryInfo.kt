package org.jetbrains.kotlin

import org.gradle.api.Project

@OptIn(ExperimentalStdlibApi::class)
data class EndorsedLibraryInfo(konst project: Project, konst name: String) {

    konst projectName: String
        get() = project.name

    konst taskName: String by lazy {
        projectName.split('.').joinToString(separator = "") { name -> name.replaceFirstChar { it.uppercase() } }
    }
}