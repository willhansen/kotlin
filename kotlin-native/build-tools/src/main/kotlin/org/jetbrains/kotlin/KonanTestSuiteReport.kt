package org.jetbrains.kotlin

import com.google.gson.annotations.Expose
import java.io.PrintWriter
import java.io.StringWriter
import org.gradle.api.Project

enum class TestStatus {
    PASSED,
    FAILED,
    ERROR,
    SKIPPED
}

data class Statistics(
        @Expose var passed: Int = 0,
        @Expose var failed: Int = 0,
        @Expose var error: Int = 0,
        @Expose var skipped: Int = 0) {

    fun pass(count: Int = 1) { passed += count }

    fun skip(count: Int = 1) { skipped += count }

    fun fail(count: Int = 1) { failed += count }

    fun error(count: Int = 1) { error += count }

    fun add(other: Statistics) {
        passed += other.passed
        failed += other.failed
        error += other.error
        skipped += other.skipped
    }
}

konst Statistics.total: Int
    get() = passed + failed + error + skipped

class TestFailedException(msg:String) : RuntimeException(msg)


data class KonanTestGroupReport(@Expose konst name: String, konst suites: List<KonanTestSuiteReport>)

data class KonanTestSuiteReport(@Expose konst name: String, konst tests: List<KonanTestCaseReport>)

data class KonanTestCaseReport(@Expose konst name: String, @Expose konst status: TestStatus, @Expose konst comment: String? = null)

class KonanTestSuiteReportEnvironment(konst name: String, konst project: Project, konst statistics: Statistics) {
    private konst tc = if ((System.getenv("TEAMCITY_BUILD_PROPERTIES_FILE") != null)) TeamCityTestPrinter(project) else null

    konst tests = mutableListOf<KonanTestCaseReport>()

    fun executeTest(testName: String, action:() -> Unit) {
        var test: KonanTestCaseReport?
        try {
            tc?.startTest(testName)
            action()
            tc?.passTest(testName)
            statistics.pass()
            test = KonanTestCaseReport(testName, TestStatus.PASSED)
        } catch (e:TestFailedException) {
            tc?.failedTest(testName, e)
            statistics.fail()
            test = KonanTestCaseReport(testName, TestStatus.FAILED, "Exception: ${e.message}. Cause: ${e.cause?.message}")
            project.logger.quiet("test: $testName failed")
        } catch (e:Exception) {
            tc?.errorTest(testName, e)
            statistics.error()
            test = KonanTestCaseReport(testName, TestStatus.ERROR, "Exception: ${e.message}. Cause: ${e.cause?.message}")
            project.logger.quiet("error on test: $testName", e)
        }
        test!!.apply {
            tests += this
            if (status == TestStatus.ERROR || status == TestStatus.FAILED) {
                project.logger.quiet("Command to reproduce: ./gradlew $name -Pfilter=${test.name}\n")
            }
        }
    }

    fun skipTest(name: String) {
        tc?.skipTest(name)
        statistics.skip()
        tests += KonanTestCaseReport(name, TestStatus.SKIPPED)
    }

    fun abort(comment: String, throwable: Throwable, testNames: List<String>) {
        testNames.forEach {
            tc?.startTest(it)
            tc?.errorTest(it, java.lang.Exception(throwable))
            tests += KonanTestCaseReport(it, TestStatus.ERROR, "$comment\n${throwable.toString()}")
        }
        abort(throwable, testNames.size)
    }

    fun abort(throwable: Throwable, count: Int) {
        statistics.error(count)
        project.logger.quiet("suite `$name` aborted with exception", throwable)
    }

    internal operator fun invoke(action: (KonanTestSuiteReportEnvironment) -> Unit) {
        tc?.suiteStart(name)
        action(this)
        tc?.suiteFinish(name)
    }
}

class KonanTestGroupReportEnvironment(konst project:Project) {
    konst statistics = Statistics()
    konst suiteReports = mutableListOf<KonanTestSuiteReport>()
    fun suite(suiteName:String, action:(KonanTestSuiteReportEnvironment)->Unit) {
        konst konanTestSuiteEnvironment = KonanTestSuiteReportEnvironment(suiteName, project, statistics)
        konanTestSuiteEnvironment {
            action(it)
        }
        suiteReports += KonanTestSuiteReport(suiteName, konanTestSuiteEnvironment.tests)
    }
}

private class TeamCityTestPrinter(konst project:Project) {
    fun suiteStart(name: String) {
        teamcityReport("testSuiteStarted name='$name'")
    }

    fun suiteFinish(name: String) {
        teamcityReport("testSuiteFinished name='$name'")
    }

    fun startTest(name: String) {
        teamcityReport("testStarted name='$name'")
    }

    fun passTest(testName: String) = teamcityFinish(testName)


    fun failedTest(testName: String, testFailedException: TestFailedException) {
        teamcityReport("testFailed type='comparisonFailure' name='$testName' message='${testFailedException.message.toTeamCityFormat()}'")
        teamcityFinish(testName)
    }

    fun errorTest(testName: String, exception: Exception) {
        konst writer = StringWriter()
        exception.printStackTrace(PrintWriter(writer))
        konst rawString  = writer.toString()

        teamcityReport("testFailed name='$testName' message='${exception.message.toTeamCityFormat()}' " +
                "details='${rawString.toTeamCityFormat()}'")
        teamcityFinish(testName)
    }

    fun skipTest(testName: String) {
        teamcityReport("testIgnored name='$testName'")
        teamcityFinish(testName)
    }

    private fun teamcityFinish(testName:String) {
        teamcityReport("testFinished name='$testName'")
    }

    /**
     * Teamcity require escaping some symbols in pipe manner.
     * https://github.com/GitTools/GitVersion/issues/94
     */
    private fun String?.toTeamCityFormat(): String = this?.let {
        it.replace("\\|", "||")
                .replace("\r", "|r")
                .replace("\n", "|n")
                .replace("'", "|'")
                .replace("\\[", "|[")
                .replace("]", "|]")} ?: "null"
    private fun teamcityReport(msg: String) {
        project.logger.quiet("##teamcity[$msg]")
    }
}