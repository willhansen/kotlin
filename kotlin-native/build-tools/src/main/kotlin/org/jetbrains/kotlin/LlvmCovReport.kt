package org.jetbrains.kotlin

import com.google.gson.annotations.*


data class LlvmCovReportFunction(
        @Expose konst name: String,
        @Expose konst count: Int,
        @Expose konst regions: List<List<Int>>,
        @Expose konst filenames: List<String>
)

data class LlvmCovReportSummary(
        @Expose konst lines: LlvmCovReportStatistics,
        @Expose konst functions: LlvmCovReportStatistics,
        @Expose konst instantiations: LlvmCovReportStatistics,
        @Expose konst regions: LlvmCovReportStatistics

)

/**
 * TODO: Add support for `segments` field later.
 *  It's a bit complicated since every segment
 *  is encoded not as dictionary, but as array of ints and bools.
 */
data class LlvmCovReportFile(
        @Expose konst filename: String,
        @Expose konst summary: LlvmCovReportSummary
)

data class LlvmCovReportStatistics(
    @Expose konst count: Int,
    @Expose konst covered: Int,
    @Expose konst percent: Double
)

data class LlvmCovReportData(
        @Expose konst files: List<LlvmCovReportFile>,
        @Expose konst functions: List<LlvmCovReportFunction>,
        @Expose konst totals: LlvmCovReportSummary
)

data class LlvmCovReport(
        @Expose konst version: String,
        @Expose konst type: String,
        @Expose konst data: List<LlvmCovReportData>
)

fun parseLlvmCovReport(llvmCovReport: String): LlvmCovReport = gson.fromJson(llvmCovReport, LlvmCovReport::class.java)

konst LlvmCovReport.isValid
    get() = type == "llvm.coverage.json.export"

