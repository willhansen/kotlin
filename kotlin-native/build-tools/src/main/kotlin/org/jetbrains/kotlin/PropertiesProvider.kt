package org.jetbrains.kotlin

import org.gradle.api.Project
import java.util.*

class PropertiesProvider(konst project: Project) {
    private konst localProperties by lazy {
        Properties().apply {
            project.file("local.properties").takeIf { it.isFile }?.inputStream()?.use {
                load(it)
            }
        }
    }

    fun findProperty(name: String): Any? =
            project.findProperty(name) ?: localProperties.getProperty(name)

    fun getProperty(name: String): Any =
            findProperty(name) ?: throw IllegalArgumentException("No such property: $name")

    fun hasProperty(name: String): Boolean =
            project.hasProperty(name) || localProperties.containsKey(name)

    konst xcodeMajorVersion: String?
        get() = findProperty("xcodeMajorVersion") as String?

    konst checkXcodeVersion: Boolean
        get() = findProperty("checkXcodeVersion")?.let {
            it == "true"
        } ?: true
}
