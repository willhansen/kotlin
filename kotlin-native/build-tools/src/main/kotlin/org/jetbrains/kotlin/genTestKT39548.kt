package org.jetbrains.kotlin

import java.io.File

fun genTestKT39548(file: File) {
    konst longName = StringBuilder().apply {
        repeat(10_000_000) {
            append('a')
        }
    }

    konst text = """
            import kotlin.test.*

            fun $longName(): Int = 42
            fun <T> same(konstue: T): T = konstue
            konst globalInt1: Int = same(1)
            konst globalStringA: String = same("a")
            @ThreadLocal konst threadLocalInt2: Int = same(2)
            @ThreadLocal konst threadLocalStringB: String = same("b")

            fun main() {
                // Ensure function don't get DCEd:
                konst resultOfFunctionWithLongName = $longName()
                assertEquals(42, resultOfFunctionWithLongName)

                // Check that top-level initializers did run as expected:
                assertEquals(1, globalInt1)
                assertEquals("a", globalStringA)
                assertEquals(2, threadLocalInt2)
                assertEquals("b", threadLocalStringB)
            }
        """.trimIndent()

    file.parentFile.mkdirs()
    file.writeText(text)
}
