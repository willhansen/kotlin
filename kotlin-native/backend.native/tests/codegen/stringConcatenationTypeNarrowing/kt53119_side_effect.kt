/*
 * Copyright 2010-2022 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */

// The same tests are also tested with FileCheck in folder ../../filecheck/kt53119_*.kt
// Please keep them in sync

package codegen.stringConcatenationTypeNarrowing.kt53119_side_effect
import kotlin.test.*

fun withSideEffect(): String? {
    println("Side effect")
    return "Result."
}

@Test
fun runTest() {
    println("Retkonst is: " + withSideEffect())
    println("Retkonst is: " + withSideEffect() + "!")
    println(withSideEffect() + withSideEffect())
    println(withSideEffect() + withSideEffect() + withSideEffect())
}
