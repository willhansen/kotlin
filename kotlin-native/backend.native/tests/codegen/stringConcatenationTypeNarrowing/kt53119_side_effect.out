Side effect
Retkonst is: Result.
Side effect
Retkonst is: Result.!
Side effect
Side effect
Result.Result.
Side effect
Side effect
Side effect
Result.Result.Result.
